include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,435 @@
+
+        /*
+         * Initial volume of the phase that is enforced by the Lagrange multiplier
+         */
+        double d_vol_init = std::numeric_limits<double>::quiet_NaN();
+
+        /*
+         * Target volume of the phase that is enforced by the Lagrange multiplier
+         */
+        double d_vol_target = std::numeric_limits<double>::quiet_NaN();
+
+        /*
+         * The Lagrange multiplier which enforces volume conservation.
+         */
+        double d_q = std::numeric_limits<double>::quiet_NaN();
+
+        double d_time = 0.0;
+
+        int d_interval = 1, d_max_its = 4;
+
+        double d_rel_tol = 1e-12;
+
+        void getFromRestart();
+
+        void getFromInput(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db);
+    };
+
+    /*!
+     * \brief Compute the value of the Lagrange multiplier \f$ q \f$ and use that to adjust the flow level set variable
+     * \f$ \tilde{\phi} \f$ to satisfy the constraint: \f$ f(q) = \int_{\Omega} H(\tilde{\phi} + q) \text{d}\Omega - V^0
+     * = 0 \f$. where, \f$ \tilde{\phi} \f$ is the level set field obtained from the reinitialization procedure, \f$ V^0
+     * \f$ is the volume of the fluid at \f$ t = 0 \f$ s which needs to be conserved. Here, \f$ q \f$ is computed using
+     * the Newton's method till required tolerance. In practice the level set mass loss would be fixed during the
+     * postprocess integrate hierarchy stage. Hence the application time would be the new time and the variable context
+     * would be the new context.
+     */
+    void fixMassLoss2PhaseFlows(double current_time,
+                                double new_time,
+                                bool skip_synchronize_new_state_data,
+                                int num_cycles,
+                                void* ctx);
+
+    /*!
+     * \brief Compute the value of the Lagrange multiplier \f$ q \f$ and use that to adjust the flow level set \f$
+     * \tilde{\phi \f$ when there is a solid phase in the domain with level set \f$ \psi < 0 \f$. Satisfies the
+     * constraint: \f$ f(q) = \int_{\Omega} H(\tilde{\phi} + q) H(\psi) \text{d}\Omega - V^0 = 0 \f$. where, \f$
+     * \tilde{\phi} \f$ is the level set field obtained from the reinitialization procedure, \f$ V^0 \f$ is the volume
+     * of the fluid at \f$ t = 0 \f$ s which needs to be conserved. Here, \f$ q \f$ is computed using the Newton's
+     * method till required tolerance. In practice the level set mass loss would be fixed during the postprocess
+     * integrate hierarchy stage. Hence the application time would be the new time and the variable context would be the
+     * new context.
+     */
+    void fixMassLoss3PhaseFlows(double current_time,
+                                double new_time,
+                                bool skip_synchronize_new_state_data,
+                                int num_cycles,
+                                void* ctx);
```

Comment:
Note in the documentations that `ctx` must be of type `LevelSetMassLossFixer`

Reply:
If `ctx` should have a particular type, can that be encoded in the function signature? This would be better than relying on code documentation.